### ISSUE TITLE
[flang][coarray] Unexpected error: 'llvm.mlir.addressof' op must reference a global defined by 'llvm.mlir.global', 'llvm.mlir.alias' or 'llvm.func' or 'llvm.mlir.ifunc'

### ISSUE THREAD

**Issue by sscalpone** [llvm/llvm-project#193157]
Fujitsu Fortran test 0907/0907_0015.f90 fails

Here's a smaller reproducer:

```fortran
block
integer :: kk
allocatable :: kk(:)[:]
endblock
end
```

```
% flang -fcoarray -c 0907/0907_0015.f90
warning: Support for multi image Fortran features is still experimental and in development.
loc("0907/0907_0015.f90":1:1): error: 'llvm.mlir.addressof' op must reference a global defined by 'llvm.mlir.global', 'llvm.mlir.alias' or 'llvm.func' or 'llvm.mlir.ifunc'
error: Lowering to LLVM IR failed
flang: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:782: llvm::LogicalResult convertOperationImpl(mlir::Operation&, llvm::IRBuilderBase&, mlir::LLVM::ModuleTranslation&): Assertion `(global || function || alias || ifunc) && "referencing an undefined global, function, alias, or ifunc"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and instructions to reproduce the bug.
Stack dump:
0.      Program arguments: bin/flang -fc1 -triple x86_64-unknown-linux-gnu -emit-obj -fcolor-diagnostics -mrelocation-model pic -pic-level 2 -pic-is-pie -target-cpu x86-64 -fcoarray -resource-dir lib/clang/23 -mframe-pointer=all -o 0907_0015.o -x f95 0907/0907_0015.f90
 #0 0x0000000006636ce8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (bin/flang+0x6636ce8)
 #1 0x0000000006633b14 llvm::sys::RunSignalHandlers() (bin/flang+0x6633b14)
 #2 0x0000000006633c8f SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007f226b63fc30 __restore_rt (/lib64/libc.so.6+0x3fc30)
 #4 0x00007f226b68d02c __pthread_kill_implementation (/lib64/libc.so.6+0x8d02c)
 #5 0x00007f226b63fb86 gsignal (/lib64/libc.so.6+0x3fb86)
 #6 0x00007f226b629873 abort (/lib64/libc.so.6+0x29873)
 #7 0x00007f226b62979b _nl_load_domain.cold (/lib64/libc.so.6+0x2979b)
 #8 0x00007f226b6388c6 (/lib64/libc.so.6+0x388c6)
 #9 0x0000000008f68599 convertOperationImpl(mlir::Operation&, llvm::IRBuilderBase&, mlir::LLVM::ModuleTranslation&) LLVMToLLVMIRTranslation.cpp:0:0
#10 0x000000000aea983e mlir::LLVM::ModuleTranslation::convertOperationImpl(mlir::Operation&, llvm::IRBuilderBase&, bool) (bin/flang+0xaea983e)
#11 0x000000000aeacd17 mlir::LLVM::ModuleTranslation::convertBlockImpl(mlir::Block&, bool, llvm::IRBuilderBase&, bool) (bin/flang+0xaeacd17)
#12 0x000000000aeb3723 mlir::LLVM::ModuleTranslation::convertOneFunction(mlir::LLVM::LLVMFuncOp) (bin/flang+0xaeb3723)
#13 0x000000000aeb49a4 mlir::LLVM::ModuleTranslation::convertFunctions() (bin/flang+0xaeb49a4)
#14 0x000000000aec164a mlir::translateModuleToLLVMIR(mlir::Operation*, llvm::LLVMContext&, llvm::StringRef, bool) (bin/flang+0xaec164a)
#15 0x000000000669bb67 Fortran::frontend::CodeGenAction::generateLLVMIR() (bin/flang+0x669bb67)
#16 0x00000000066a0218 Fortran::frontend::CodeGenAction::executeAction() (bin/flang+0x66a0218)
#17 0x000000000667fc9d Fortran::frontend::FrontendAction::execute() (bin/flang+0x667fc9d)
#18 0x0000000006662f3b Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) (bin/flang+0x6662f3b)
#19 0x000000000668abec Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) (bin/flang+0x668abec)
#20 0x0000000005ed3761 fc1_main(llvm::ArrayRef<char const*>, char const*) (bin/flang+0x5ed3761)
#21 0x0000000005ed302e main (bin/flang+0x5ed302e)
#22 0x00007f226b62a610 __libc_start_call_main (/lib64/libc.so.6+0x2a610)
#23 0x00007f226b62a6c0 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2a6c0)
#24 0x0000000005ed0eee _start (bin/flang+0x5ed0eee)
flang-23: error: unable to execute command: Aborted (core dumped)
flang-23: error: flang frontend command failed with exit code -2 (use -v to see invocation)
flang version 23.0.0git (https://github.com/llvm/llvm-project 1bfcddcdedb0ca7095d7c4475ec8b6a641f84d3f)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: bin
Build config: +assertions
flang-23: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
flang-23: note: diagnostic msg: /tmp/0907_0015-e5d97f
flang-23: note: diagnostic msg: /tmp/0907_0015-e5d97f.sh
flang-23: note: diagnostic msg:

********************
```

**Comment by bonachea** [comment]
Fixed in #213890

**Comment by bonachea** [comment]
Backport #214321
